### DISCUSSION TITLE
Feature Proposal: Adding DateTime Formatting Support

### DISCUSSION THREAD

**Discussion by harshdoesdev** [fastn-stack/fastn#1039] [Ideas & RFCs]
**Summary:**
Hey everyone! I've been thinking about a cool feature we could add to our framework. It's all about Date Time formatting. So, imagine we fetch a timestamp or date from an SQL database, but we want to display it in a specific format. That's where this feature comes in handy!

**Proposal:**
I suggest we introduce a new data type, let's call it `ftd.datetime`, that handles datetime values and formatting. This way, we can easily format the fetched datetime values consistently throughout our framework.

**Integration with Chrono:**
To make this happen, we'll need to bring in a library called `chrono`. It's super helpful for handling datetime formatting operations, so we won't have to reinvent the wheel.

**Parsing and Handlers:**
We'll need to tweak our parser to recognize the `ftd.datetime` data type and its syntax. And of course, we'll also write some neat handlers (functions) that take care of formatting the datetime values and passing them to other components, like `ftd.text`, for rendering.

**Storing and Passing Unix Timestamps:**
To keep things standardized, we can store and pass the datetime values internally as Unix timestamps. We'll use the timestamps in seconds whenever we need to work with datetime values.

**Frontend and Backend Considerations:**
Remember, we have to handle datetime operations both in the frontend (JavaScript) and backend. So, we'll ensure that conversions between integer timestamps and timezones or dates are done correctly on both sides.

**Benefits:**

- Consistent datetime formatting: With ftd.datetime, we'll have a reliable way to format datetime values consistently.
Customizable output: Users can easily define the desired format using format templates.
- Less code changes: By leveraging `chrono`, we can minimize the modifications required in our existing code.
- Better organization: Having a dedicated data type for datetime values will make our code cleaner and easier to maintain.
- Backend-Frontend harmony: Ensuring datetime operations are handled correctly on both sides will lead to consistent results.

**Challenges:**
- Comprehensive datetime modeling: We might need to think a bit more about how we model datetime values internally, especially if we encounter limitations with Chrono.
- Conversions between datetime representations: Making sure we handle conversions between Unix timestamps and different timezones or dates correctly can be a bit tricky. We'll need to handle it with care.

In a nutshell, this feature will empower us to format datetime values like experts. Let's dive in, have a discussion, and fine-tune the concept together!

**Comment by amitu** [comment]
We will have to figure out date time parsing helpers for converting human written date times to internal representation. Can you write some notes on what kind of formats we should support. 

We have to worry about time zones also. Maybe we add some sort of default timezone in FASTN.ftd so people can use omit it and get concise representations. 

And also note some of the formatting functions, to convert internal representation to human readable strings

We will also need extractors, like fine the year of a give datetime or week number etc. 

We will also have to support timedelta type to capture difference between dates. We will also need type to represent datetime windows. 

This is def a big feature which will require a lot of thought.

**Reply by harshdoesdev** [reply]
Sure, I was just working on that.

**Comment by harshdoesdev** [comment]
1. Keeping the datetime value as a UNIX Epoch timestamp should solve a lot of problems, such as timedelta, conversion, etc.
2. `chrono` and `chrono-tz` can handle both formatting and timezones.
3. Allowing the user to define the timezone of their choice in the FASTN.ftd file is a good idea. They can then override the default timezone settings if they want to.
4. We can use a fixed format (ISO Date Format) for human-written datetimes, and ISO dates are fairly easy to parse. The format should be `YYYY-MM-DDTHH:mm:ss.sssZ`.
5. Instead of adding methods to the DateTime type, we can create a new `processor` for formatting and handling datetimes. This approach is more efficient and functional because it can be used by different components without requiring significant code changes.

Example:
```
-- import: fastn/processors as pr

-- datetime today:
$processor$: pr.datetime
format: YYYY-MM-DD

-- ftd.text: $today
```

A Basic Implementation of DateTime Type could be something like this:

```rust
struct DateTime {
    value: i64, // Integer timestamp representing Unix epoch
}

impl DateTime {
    fn new(value: i64) -> DateTime {
        DateTime { value }
    }
}

fn create_datetime() -> DateTime {
    // Implementation for creating a DateTime instance (similar to `new()`)
    // goes here
}

fn get_current_datetime() -> DateTime {
    // Implementation for getting the current datetime
    // goes here
}

fn format_datetime(datetime: &DateTime, string: &str) {
    // Formatting logic for datetime
}

fn calculate_timedelta(datetime1: &DateTime, datetime2: &DateTime) -> i64 {
    return datetime1.value - datetime2.value
}

fn datetime_to_string(datetime: &DateTime) -> String {
    // Conversion logic for datetime to a human-readable format
    // We can use the chrono crate for formatting the timestamp
    // Example:
    // chrono::NaiveDateTime::from_timestamp(datetime.value, 0).to_string()
}
```

**Comment by amitu** [comment]
When discussing language features do not bother with implementation. Focus on use cases and show syntax and code snippets that user will write. 

Compile a list of uses cases and corresponding code snippets for datetime.

**Comment by harshdoesdev** [comment]
Some common use cases are:

## 1. Fetching data from a SQL database

**Current**
```
-- import: fastn/processors as pr

-- record user:
integer id:
string name:
string department:
string date_created:

-- user u:
$processor$: pr.pg
id: 1

SELECT * FROM "users" where id = $id ;

-- ftd.text: $u.date_created
```

**Proposed**
```
-- import: fastn/processors as pr

-- record user:
integer id:
string name:
string department:
datetime date_created:

-- user u:
$processor$: pr.pg
id: 1

SELECT * FROM "users" where id = $id ;
```

Now, the datetime formatting can be done in two ways:

### Using Processor

```
-- ftd.text: $u.date_created
$processor$: pr.datetime
format: YYYY-MM-DD
```

**Benefits of using a processor**
- The implementation becomes very easy because datetime is just another integer.
- It is clearer and adheres to FP principles

### Using method

```
-- ftd.text: $u.date_created.format(YYYY-MM-DD)
```

**Reply by amitu** [reply]
```
-- ftd.text: $u.date_created.format(YYYY-MM-DD)
```

We don't do that. We have a ftd.integer to render integer values, a ftd.boolean for rendering boolean and so on. Converting things to string to show to user is bad idea as it does not allow people to then style that string, what if I want to let people hover over each part, say year, date month, and show some tooltip, like number of events in that year month etc. The convert to string erases all information, and you are left with just a string.

**Comment by amitu** [comment]
@harshdoesdev please keep all implementation related concerns "easy to implement" and "we will use chrono" out of this discussion. We are creating a language that billions of people will *love*, we do not worry too much about easy to implement, it must be beautiful, it must be something we are proud of etc.

**Reply by harshdoesdev** [reply]
I will try to focus more on the feature aspect rather than the implementation part. It was my first time writing such a proposal, so please bear with me.

**Comment by harshdoesdev** [comment]
@amitu Sure enough, I had a discussion about this with @Arpita-Jaiswal  as well, which helped clear up many doubts. Now, I am working on defining a new syntax and flow for handling date times. Dealing with date and time can pose significant challenges in almost every programming language, mainly due to the lack of sufficient emphasis from the start. In improving the proposals, I will consider the following aspects:

- Easy syntax
- Cross-platform support
- Timezone support
- Value conversion
- Uniform syntax for both client-side and server-side code
- How the syntax will be hydrated once execution is transferred to the browser.
